vertx-pojongo/src/main/java/de/braintags/vertx/jomnigate/mongo/dataaccess/MongoWrite.java

```diff
@@ -192,28 +199,60 @@ private void handleInsertError(Throwable t, T entity, MongoStoreObject<T> storeO
     }
   }
 
-  private void doUpdate(T entity, MongoStoreObject<T> storeObject, Handler<AsyncResult<IWriteEntry>> resultHandler) {
+  private void doUpdate(final T entity, final MongoStoreObject<T> storeObject,
+      final Handler<AsyncResult<IWriteEntry>> resultHandler) {
     MongoClient mongoClient = (MongoClient) ((MongoDataStore) getDataStore()).getClient();
     IMapper<T> mapper = getMapper();
     String collection = mapper.getTableInfo().getName();
     final Object currentId = storeObject.get(mapper.getIdInfo().getField());
-    String idFieldName = mapper.getIdInfo().getField().getColumnInfo().getName();
-    JsonObject query = new JsonObject();
-    query.put(idFieldName, currentId);
 
-    mongoClient.save(collection, storeObject.getContainer(), result -> {
-      if (result.failed()) {
-        resultHandler.handle(Future.failedFuture(new WriteException(result.cause())));
-      } else {
-        LOG.debug("updated");
-        finishUpdate(currentId, entity, storeObject, resultHandler);
-      }
-    });
+    if (getQuery() != null) {
+      IQuery<T> q = getDataStore().createQuery(getMapperClass());
+      q.setSearchCondition(ISearchCondition.and(ISearchCondition.in(mapper.getIdInfo().getIndexedField(), currentId),
+          getQuery().getSearchCondition()));
+      q.buildQueryExpression(null, queryExpRes -> {
+        mongoClient.replaceDocuments(collection, ((MongoQueryExpression) queryExpRes.result()).getQueryDefinition(),
+            storeObject.getContainer(), res -> {
+              if (res.succeeded()) {
+                finishQueryUpdate(currentId, entity, storeObject, res.result(), resultHandler);
+              } else {
+                resultHandler.handle(Future.failedFuture(new WriteException(res.cause())));
+              }
+            });
+      });
+    } else {
+      mongoClient.save(collection, storeObject.getContainer(), result -> {
+        if (result.failed()) {
+          resultHandler.handle(Future.failedFuture(new WriteException(result.cause())));
+        } else {
+          LOG.debug("updated");
+          finishUpdate(currentId, entity, storeObject, resultHandler);
+        }
+      });
+    }
+
   }
 
-  @SuppressWarnings({ "unchecked", "rawtypes" })
-  private void finishInsert(Object id, T entity, MongoStoreObject storeObject,
-      Handler<AsyncResult<IWriteEntry>> resultHandler) {
+  private void finishQueryUpdate(final Object id, final T entity, final MongoStoreObject<T> storeObject,
+      final MongoClientUpdateResult updateResult, final Handler<AsyncResult<IWriteEntry>> resultHandler) {
+    if (updateResult.getDocMatched() != 0 && updateResult.getDocMatched() == updateResult.getDocModified()) {
+      executePostSave(entity, lcr -> {
```

Comment:
Wouldn't it be better to call finishUpdate here instead of duplicating its code?